src/v/storage/segment_reader.h

```diff
@@ -26,20 +29,88 @@
 
 namespace storage {
 
+class segment_reader;
+
+/**
+ * A segment reader handle confers the right to use a file descriptor
+ * opened by segment_reader, a clone of which is encapsulated in
+ * the handler's input_stream (if stream is set).
+ *
+ * We need this handle & associated reference counting to enable
+ * opening segment files on demand, and keeping them open as long
+ * as input_streams using the file are alive.
+ */
+class segment_reader_handle {
```

Comment:
how much pre-requisites can we introduce in a previous commit? i'm finding this commit hard to work through there is a lot of stuff changing but it also looks like quite a bit is greenfield and could come before?